mitodl/open-discussions
# Update ContentFile (Resource) ES doc structure

#### Pre-Flight checklist

#### What are the relevant tickets?
Closes #3131 

#### What's this PR do?
- Adds `slug` field to `LearningResourceRun` model, adds `slug` to the ES LearningResourceRun docs.
- Uses the S3 url as the default url for all non-page resource files except external (non-OCW) links.
- Makes resource page url's relative.
- Populates/adds `topics`, `run_slug`, `short_url` and `run_title` to `ContentFile` (resource) docs in ElasticSearch.
- Adds 2 new `content_type` values, `pdf` and `video`, determined from the `file_type` (ie mime type)


#### How should this be manually tested?
- Run the new migration.  All OCW LearningResourceRuns should have a populated `slug` field afterward.
- Recreate the index.
- Run `backpopulate_ocw_files` management command.  
- Perform a search for OCW courses.  The raw search results should include a populated `slug` attribute for each course run.
- Run this search via curl/postman for resources, tweaking the content_type (try `pdf`, `page`, `video`, `file`).  Each resource should have the following populated fields: `topics`, `run_title`, `run_slug`, `content_type`, `file_type`.  The content_type should be `pdf` if the `file_type` ends with `pdf`, or `video` if the `file_type` starts with `video/`, or `page` if the `file_type` ends with `Section`, or `file` otherwise.

```
{  "size": 100,
    "post_filter": {
        "bool": {
            "must": [
                {
                    "bool": {
                        "should": [
                            {
                                "term": {
                                    "object_type.keyword": "resourcefile"
                                }
                            }
                        ],
                        "must": [
                                {"term": {
                                    "content_type": "pdf"
                                }   }                         
                        ]
                    }
                },
                {
                    "bool": {
                        "should": [
                            {
                                "has_parent": {
                                    "parent_type": "resource",
                                    "query": {
                                        "bool": {
                                            "should": [
                                                {
                                                    "term": {
                                                        "offered_by": "OCW"
                                                    }
                                                }
                                            ]
                                        }
                                    }
                                }
                            }
                        ]
                    }
                }
            ]
        }
    }
}
```
